pages/developers.tsx

```diff
@@ -37,49 +49,43 @@ const options_time = [
 export default function PageDevelopers() {
   const {dispatch, state: {loading}} = useContext(ApplicationContext);
   const [issues, setIssues] = useState<IssueData[]>([]);
-  const [filterStateIssues, setfilterStateIssues] = useState({
-    state: "",
-    issues: [],
-  });
-
+  const [filterByState, setFilterByState] = useState<Filter>(filtersByIssueState[0]);
 
-  function handleChangeFilterIssue(params: { value: string; label: string; }) {
-    if (params.value === "all") {
-      setfilterStateIssues({
-        state: params.value,
-        issues,
-      });
-    } else {
-      setfilterStateIssues({
-        state: params.value,
-        issues: issues.filter(
-          (obj: IssueData) => obj.state.toLowerCase() === params.value
-        ),
-      });
-    }
+  function handleChangeFilterByState(filter: Filter) {
+    setFilterByState(filter);
   }
 
   function updateIssuesList(issues: IssueData[]) {
     console.log(`got issues`, issues);
     setIssues(issues);
-    if (filterStateIssues.issues.length === 0) {
-      setfilterStateIssues({ state: "all", issues });
-    }
   }
 
-  function getIssues() {
-    dispatch(changeLoadState(true))
-    GithubMicroService.getIssues()
-                      .then(updateIssuesList)
-                      .catch((error) => {
-                        console.log('Error', error)
-                      })
-                      .finally(() => {
-                        dispatch(changeLoadState(false))
-                      });
-  }
+  useEffect(() => {
```

Comment:
Please, don't hide async effects inside a sync effect function with a closure. useEffects priority is too high to have its functions hidden away in a closure.

_if you don't need to call that function anywhere else, then simply don't state a function and instead write it as direct callback_

But we rather have the function stated _outside_ of the effect and then used inside the effect callback; If nothing else, for readability and understanding sake

Reply:
I tend to follow the guidelines of Dan Abramov, who wrote useEffect.

https://overreacted.io/a-complete-guide-to-useeffect/#moving-functions-inside-effects

So maybe I need to check if there are any dependencies missing in the dependency array, but asynchronous functions can still be inside useEffect, can't they?

Reply:
> but asynchronous functions can still be inside useEffect, can't they?

And here is the crux of the thing. **YOU CAN**, but there is an error deep in useEffect that will throw an error if it catches a state update when the component is no longer mounted. While highly improbable, _you can_ catch that error.

I usually hit this error when I give in an async function as a parameter to `useEffect`, such as

```
async function test() { return Promise.resolve(1) }
useEffect(test, [])
```

This happens because `async function` returns a Promise, while `useEffect` is expecting a "unmount function" to be returned on the use effect -- thus I made it a rule to never provide `async function` inside (or anywhere near) a `useEffect` call.

Following the same rule, I'm going "against it" if I "hide" an async call inside the useEffect. While it _should not_ throw the error because it _does run_ on the correct scope, _readers_ might think otherwise - and I've been favouring readability over code-golf lately.